cli/src/utils.ts

```diff
@@ -233,3 +233,75 @@ export class Batcher<T = unknown> {
     this.items = [];
   }
 }
+
+interface HashCacheEntry {
+  hash: string;
+  mtime: number;
+  size: number;
+}
+
+interface HashCache {
+  [filepath: string]: HashCacheEntry;
+}
+
+export class FileHashCache {
+  private cache: HashCache = {};
+  private cacheFile: string;
+  private isDirty = false;
+
+  constructor(configDir: string) {
+    this.cacheFile = join(configDir, 'hash-cache.json');
+  }
+
+  async load() {
+    if (existsSync(this.cacheFile)) {
+      try {
+        const content = await readFile(this.cacheFile, 'utf-8');
+        this.cache = JSON.parse(content);
+      } catch (error) {
+        console.warn('Failed to load hash cache, starting fresh');
+        this.cache = {};
+      }
+    }
+  }
+
+  async save() {
+    if (this.isDirty) {
+      try {
+        await writeFile(this.cacheFile, JSON.stringify(this.cache, null, 2));
```

Comment:
this file is going to be significantly bigger with pretty-print, and it's not meant to be read by humans, so I would probably avoid stringifying it.